Inject friends with all RemoteMessage data

HankieCodes opened the pull request:
Closes #1 

- [x] Added `Debug()` Logging for clearer Chrome tools
- [x] Session disconnnects handlers while processing events
- [x] Improved separation of message types for port/peerjs

![image](https://user-images.githubusercontent.com/3521186/81639508-be6d9a00-93e1-11ea-93b2-895300bdcd80.png)

The screenshot above shows the Chrome console logs with highlighting for the different peers.

1. In the Red box, we see the first peer to join the session. Here's each line explainend:
    1. Blue: ready to send outgoing connection
    1. Blue: sending outgoing connection
    1. Pink: receiving device writes `this.#friends` based on `conn.metadata`.
    1. Blue: sending device cannot write `this.#friends` because they don't know the metadata of the device they connected to.
    1. Pink: receiving device waited 500ms, polled their video, and then appended `message.friends` data and sent a message
    1. Blue: write `this.#friends` to add our host, since we now know their metadata.
1. In Purple, we see the second peer join the session. It follows basically the same flow, but slightly more complex:
    1. Green: ready to send outgoing connection
    1. Green: sending outgoing connection
    1. Blue: receiving device writes `this.#friends` based on `conn.metadata`.
    1. Green: sending device cannot write `this.#friends` because they don't know the metadata of the device they connected to.
    1. Blue: receiving device waited 500ms, polled their video, and then appended `message.friends` data and sent a message
    1. Pink: bystander member is told about this new person. schedules a connection
    1. Green: write `this.#friends` to add our host, since we now know their metadata. also discovers the Pink device and schedules a connection.
1. Below the dashed line, we see the Pink and Green server make their connection:
    1. Green: The shorter randomly-selected timeout was from Green. Connection starting.
    1. Pink: receiving device writes `this.#friends` based on `conn.metadata`.
    1. Green: sending device lost track of the `friends[i]` metadata <small>(we could improve this... 🤷)</small>. so it doesn't write to `this.#friends`
    1. Pink: receiving device waited 500ms, polled their video, and then appended `message.friends` data and sent a message
    1. Blue: nothing to do since all connections are already established.
    1. Green: write `this.#friends` to add our host, since we now know their metadata.

Similar logic applies for 4, 5, 6, ... connections. The new device will use the list of `friends` on the first video status in order to connect everyone together.

![2020-05-11 23 21 30](https://user-images.githubusercontent.com/3521186/81640416-14434180-93e4-11ea-84c6-f91260a74788.gif)

If you slow the GIF down, you can see that the new joiner (bottom-most window) very quickly grows from 0 connections to 4 connections ... but it updates the total as we learn of each person's metadata.

HankieCodes committed:
#1 - Inject friends with all RemoteMessage data

- [x] Added `Debug()` Logging for clearer Chrome tools
- [x] Session disconnnects handlers while processing events
- [x] Improved separation of message types for port/peerjs

HankieCodes commented on src/background/host.ts:
```diff
@@ -1,26 +1,87 @@
+        this.#log(`Handling ${isIncoming ? "in" : "out"} connection`, conn.metadata);
+
+        if (isIncoming) {
+            // `conn.metadata` will always be the `Friend` of the **sender**, so when we handle the outgoing request,
+            // we avoid writing ourselves as a friend (and therefore emitting events to ourselves & getting
+            // infinite-looped).
+            const friend: FriendConnected = {
+                ...(conn.metadata as Friend),
+                conn
+            };
+
+            this.#friends.set(conn.peer, friend);
+            this.onChangeFriends(this.#friends.size, 1);
```
This is where I got stuck and introduced the `Math.random()`, which isn't perfect. The random works fine as long as they don't pick the same `~40ms` window.

I'm not really even sure what the solution would be ... since we wouldn't know which connection to destroy and which one to keep. B/c if the other side doesn't make the same choice, you've then destroyed both connections, lol.

HankieCodes replied:
Got this fixed using a deterministic comparison (`id` alphanumerically)

SudharsanSukumar commented on src/background/host.ts:
```diff
@@ -1,26 +1,87 @@
+            }
+            // We need to randomize `#ensureConnections()` so that two peers don't try to connect to each other at
+            // the exact same time (immediately upon receiving the message).
+            const timer = Math.round(Math.random() * 1000);
```
I think this will end up being pretty error prone

HankieCodes replied:
Yea. In [comment here](#discussion_r423467657) I talk about this. I'm not really sure what to do about it tho.

SudharsanSukumar commented on src/background/host.ts:
```diff
@@ -1,26 +1,87 @@
+            } else if (friend) {
+                // update member data without touching `friend.conn`
+                this.#friends.set(member.id, {
+                    ...member,
```
shouldn't we do ...friend, ...member so that the latest data applies over the top of existing data?

HankieCodes replied:
Good catch

SudharsanSukumar commented on src/background/host.ts:
```diff
@@ -1,26 +1,87 @@
 
-        conn.on("data", (data: Message) => {
             this.#port.postMessage(data);
+            setTimeout(() => this.#ensureConnections(data.friends, conn), timer);
```
So any time we receive a message we ensure connections again? If we instead define how we want users to join the topology, we could avoid the random wait and everyone trying to connect to a peer who is trying to connect back

HankieCodes replied:
We basically want to keep everyone in sync all the time .. right now. Until we build a muting/host-control-only feature.

I think being out-of-sync is a pretty fatal problem, and we should course-correct as much as possible until everyone agrees. The way I have it set up ... is that any change to state will tell everyone the state as it should be. Peers who already follow that state (with 250ms leeway on `video.currentTime`) do nothing .... so it's only peers who we deem too-far-from-the-desired-state that actually do work on one of these messages.

SudharsanSukumar replied:
It seems like we might not be on the same page? If we're planning on ever supporting people being in different states, then this design will have to change to look like what I already have. I also think my design is more efficient even though it's admittedly more complex

HankieCodes replied:
I don't necessarily think we would be prevented from supporting that if we still emit video events to sync the peer list.

One thing I could imagine seeing in the UI is the play time of each person. That way, you can see if anyone has de-sync'd themself from the stream ... and we could support that pretty easily if it is the recipient who ignores a `muted` person, rather than the sender.

SudharsanSukumar commented on src/background/host.ts:
```diff
@@ -1,26 +1,87 @@
+                this.onChangeFriends(this.#friends.size, 1);
+            } else {
+                // for new people, just go connect to them (& hopefully they aren't doing the same)
+                this.connect(member.id);
```
especially for users who are trying to join the session, we'll get into a situation where they could have missed messages while establishing connections. This is why I added the step to sync state from the leader after the user has established all the connections

HankieCodes replied:
So every message contains the full state data of the video and the group members + their metadata. If they're out of sync, the next message would correct them. And since we emit a message 500ms after joining, they shouldn't ever be out of sync for very long. If they are ... any member just needs to emit an event.

If this becomes a problem, we could swap this to be a heartbeat message from every member periodically.

SudharsanSukumar commented on src/background/host.ts:
```diff
@@ -1,26 +1,87 @@
 
         conn.on("close", () => {
-            this.#friends.delete(conn);
+            this.#friends.delete(conn.peer);
```
noticed that this doesn't have the concept of a leader anymore. We need a leader for permissions. There's also the issue that when someone attempts to resync after going muted, they have no way of determining which user in the session they should sync with

HankieCodes replied:
I was thinking when we implement that, we can add other properties to the `RemoteMessageExtensions` interface, like `leaderId` and `isSharedControl` and also include those as part of the state synchronization process.

SudharsanSukumar commented on src/background/host.ts:
```diff
@@ -1,26 +1,87 @@
+
+        this.#log(`Handling ${isIncoming ? "in" : "out"} connection`, conn.metadata);
+
+        if (isIncoming) {
```
depending on who is connecting to who, wouldn't this cause some users to have missing personalData for a connection until the next poll comes?

HankieCodes replied:
Kinda. The only time we have a connection with no metadata is for the sender on a new connection. But the same message that provides all the peers of the group also includes the metadata of the connection you just made, so you'd have everyone's metadata by the time you make your 2nd and beyond connection.

And for bystanders of a join, they always get the metadata before the connection comes in and/or the metadata is included in the `conn.metadata`, and so they never have a moment of a connection without metadata.

SudharsanSukumar commented on src/background/host.ts:
```diff
@@ -29,25 +90,53 @@ export class Host {
 
         this.#peer.on("connection", (conn: Peer.DataConnection) => {
-            this.#handleConnect(conn);
+            this.#handleConnect(conn, true);
```
this is gonna cause us to spam the network while new users are joining the session. If there's state mismatch between users, clients receiving the message will not know who to follow.

HankieCodes replied:
Yea, this happens already, and I don't think we need to address that here. That kind of stuff feels like it belongs in a separate PR where we handle muting, host-control-only, etc.

SudharsanSukumar commented on src/background/host.ts:
```diff
@@ -29,25 +90,53 @@ export class Host {
-            this.id = id;
+            this.#log = Debug(`peer:${id}`);
+            this.#log("Connected to broker server");
+            this.personalData = {
```
yeah on second thought this is probably better than storing the current user in the sessionUsers

SudharsanSukumar commented on src/Message.ts:
```diff
@@ -3,14 +3,28 @@ export enum MessageType {
 }
 
-export interface VideoMessage {
+export interface Friend {
```
I think this and FriendConnection should be in Friend.ts or something so it's a little cleaner

HankieCodes replied:
Can do. I just use IntelliJ IDEA Ultimate so jumping between classes is really easy with CMD+Click.

SudharsanSukumar commented on src/Message.ts:
```diff
@@ -3,14 +3,28 @@ export enum MessageType {
+export type LocalInMessage = LocalPollMessage | LocalVideoMessage;
+export type LocalOutMessage = LocalVideoMessage;
+// A remote message is transmitted through PeerJS
+export type RemoteMessage = LocalOutMessage & RemoteMessageExtensions;
```
I think if we're going to bother differentiating local and remote messages, we should make a more clear separation between local and remote. Right now a RemoteMessage is basically anything except poll

HankieCodes replied:
But the remote message needs to always include the session metadata, regardless of message type. What I have set up here is meant to grow if/when we add more message types that can be emitted (such as `MemberDisconnect` or `MemberMuted` or something)

SudharsanSukumar commented on src/Message.ts:
```diff
@@ -3,14 +3,28 @@ export enum MessageType {
 
-export type Message = VideoMessage | PollMessage;
+export interface RemoteMessageExtensions {
+    friends: Friend[];
```
I think sending friends and trying to ensure connections based on the friends added to the message, we might get into some pretty interesting race conditions while users are connecting and disconnecting while people are interacting with the video

HankieCodes replied:
Yea, I think this again comes to splitting this work of peer-connnection-circle from the whole `isSharedControl` and `muted` features.

It's valuable work, but I don't want to have PRs that aren't mergable b/c they snowball.

SudharsanSukumar commented on src/background/host.ts:
```diff
@@ -1,26 +1,87 @@
+            return;
+        }
+
+        for (const member of friends) {
```
do we care about users that are in our local map but weren't part of the users in the message we received

HankieCodes replied:
uhm. maybe (?)

in my testing, if a user disconnects from one, they disconnect from all at pretty much the exact same time. the only case where this wouldn't happen is if an error occurs for one of their `FriendConnected` peers but not the rest.

right now we don't handle errors really at all .... and i think that's fine for now. down the road, we could be smart about `conn.on('error'` and knowing when/if to do`this.#friend.delete()` on that peer. Then anyone else emitting an event would just signal to remake that connection for the two who had an error, while everyone else just keeps on chirping.

SudharsanSukumar reviewed (changes requested):
I like how you handled personal info. I think there are some fairly common error situations we could hit that aren't currently handled. 

I also think this design will have to change once we add the concept of a session leader

HankieCodes committed:
#1 - Improved `#ensureConnection` Sync w/ alphanum. compare

- [x] Avoid double-connecting peers (both send outgoing at same time) by using a deterministic comparison
- [x] Fixed stale friend data overwriting newer member data
- [x] Fixed re-emitting of handled events by deferring handler functions
- [x] Added JSDoc code comments
- [x] Split `Friend` file for multi-file reuse
